manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,790 @@
+happens-before relationship between them. Hence, this program has a data race.
+
+\section{s:local_drf}{Local data race freedom}
+
+The OCaml memory model offers strong guarantees even for programs with data races.
+It offers what is known as \emph{local data race freedom sequential
+consistency (LDRF-SC)} guarantee. A formal definition of this property is beyond
+the scope of this manual chapter. Interested readers are encouraged to read the
+PLDI 2018 paper on \href{https://doi.org/10.1145/3192366.3192421}{Bounding Data
+Races in Space and Time}.
+
+Informally, LDRF-SC says that the data race free parts of the program remain
+sequentially consistent. That is, even if the program has data races, those
+parts of the program that are disjoint from the parts with data races are
+amenable to sequential reasoning.
+
+Consider the following snippet:
+
+\begin{caml_example*}{verbatim}
+let snippet () =
+  let c = ref 0 in
+  c := 42;
+  let a = !c in
+  a
+\end{caml_example*}
+
+Observe that "c" is a newly allocated reference. Can the read of "c" return a
+value which is not 42? Surprisingly, in the C++ and Java memory models, the
+answer is yes. With the C++ memory model, if the program has a data race, even
+in unrelated parts, then the semantics is undefined. If this snippet were
+linked with a library that had a data race, then, under the C++ memory model,
+the read may return any value. Since data races on unrelated locations can
+affect program behaviour, we say that C++ memory model is not bound in space.
+
+Unlike C++, Java memory model is bound in space. But Java memory model is not
+bound in time; data races in the future will affect the past behaviour. For
+example, consider the translation of this example to Java. We assume a prior
+definition of "Class c {int x;}" and a shared \emph{non-volatile} variable "C
+g". Now the snippet may be part of a larger program with parallel threads:
+
+\begin{verbatim}
+(* Thread 1 *)
+C c = new C();
+c.x = 42;
+a = c.x;
+g = c;
+
+(* Thread 2 *)
+g.x = 7;
+\end{verbatim}
+
+The read of "c.x" and the write of "g" in the first thread are done on separate
+memory locations. Hence, the Java memory model allows them to be reordered. As
+a result, the write in the second thread may occur before the read of "c.x",
+and hence, "c.x" returns "7".
+
+The OCaml equivalent of the Java code above is:
+
+\begin{caml_example*}{verbatim}
+let g = ref None
+
+let d1 () =
+  let c = ref 0 in
+  c := 42;
+  let a = !c in
+  g := Some c;
+  a
+
+let d2 () =
+  match !g with
+  | None -> ()
+  | Some c -> c := 7
+\end{caml_example*}
+
+Observe that there is a data race on both "g" and "c". Consider only the first
+three instructions in "d1":
+
+\begin{verbatim}
+let c = ref 0 in
+c := 42;
+let a = !c in
+...
+\end{verbatim}
+
+The OCaml memory model is bound both in space and time. The only memory location
+here is "c". Reasoning only about this snippet, there is neither the data race
```

Comment:
What about `g`? It would be much clearer if `snippet` was really a part of the larger piece of code. This is made possible if `snippet` returns `c` as proposed above.